consensus/src/data_manager.rs

```diff
@@ -4,53 +4,87 @@
 use std::sync::Arc;
 // use futures::channel::{mpsc, mpsc::Sender, oneshot};
 use crate::quorum_store::batch_reader::BatchReader;
+use aptos_crypto::HashValue;
 use aptos_types::transaction::SignedTransaction;
 use arc_swap::ArcSwapOption;
 use consensus_types::common::Payload;
 use consensus_types::proof_of_store::LogicalTime;
+use consensus_types::request_response::ConsensusRequest;
 use executor_types::Error;
+use futures::channel::mpsc::Sender;
 use tokio::sync::oneshot;
 
 /// Notification of execution committed logical time for QuorumStore to clean.
 #[async_trait::async_trait]
 pub trait DataManager: Send + Sync {
     /// Notification of committed logical time
-    async fn notify_commit(&self, logical_time: LogicalTime);
+    async fn notify_commit(&self, logical_time: LogicalTime, payloads: Vec<Payload>);
 
-    fn new_epoch(&self, data_reader: Arc<BatchReader>);
+    fn new_epoch(
+        &self,
+        data_reader: Arc<BatchReader>,
+        quorum_store_wrapper_tx: Sender<ConsensusRequest>,
+    );
 
-    async fn get_data(&self, payload: Payload) -> Result<Vec<SignedTransaction>, Error>;
+    async fn get_data(&self, maybe_payload: Payload) -> Result<Vec<SignedTransaction>, Error>;
 }
 
 /// Execution -> QuorumStore notification of commits.
 pub struct QuorumStoreDataManager {
-    data_reader: ArcSwapOption<BatchReader>, // TODO: consider arc_swap
+    data_reader: ArcSwapOption<BatchReader>,
+    quorum_store_wrapper_tx: ArcSwapOption<Sender<ConsensusRequest>>,
 }
 
 impl QuorumStoreDataManager {
     /// new
     pub fn new() -> Self {
         Self {
             data_reader: ArcSwapOption::from(None),
+            quorum_store_wrapper_tx: ArcSwapOption::from(None),
         }
     }
 }
 
 #[async_trait::async_trait]
 impl DataManager for QuorumStoreDataManager {
-    async fn notify_commit(&self, logical_time: LogicalTime) {
+    async fn notify_commit(&self, logical_time: LogicalTime, payloads: Vec<Payload>) {
         self.data_reader
             .load()
             .as_ref()
-            .unwrap() //TODO: can this be None? Need to make sure we call new_epoch() first.
+            .unwrap()
             .update_certified_round(logical_time)
             .await;
+
+        let digests: Vec<HashValue> = payloads
+            .into_iter()
+            .map(|payload| match payload {
+                Payload::DirectMempool(_) => {
+                    unreachable!()
+                }
+                Payload::InQuorumStore(proofs) => proofs,
+                Payload::Empty => Vec::new(),
+            })
+            .flatten()
+            .map(|proof| proof.digest().clone())
+            .collect();
+
+        self.quorum_store_wrapper_tx
+            .load()
+            .as_ref()
+            .unwrap()
+            .as_ref()
+            .clone()
+            .try_send(ConsensusRequest::CleanRequest(logical_time, digests))
+            .expect("could not send to wrapper");
```

Comment:
This is interesting. I think I understand now why there was confusion over CleanRequest being a ConsensusRequest.

I was originally thinking CleanRequest is sent to wrapper, then wrapper calls data_manager.notify_commit().

I guess the reasoning for doing it this way is that state_computer already has Arc<DataManager> so it can call notify_commit directly. If we do it this way, we can make CleanRequest not be a ConsensusRequest.